### ISSUE TITLE
Mouseover on an empty monitor doesn't bring focus to it.

### ISSUE THREAD

**Issue by SweepingBishops** [qtile/qtile#3381]
In a double monitor setup click on  monitor 1 (both the monitors are currently empty,i.e have no open windows). This causes the keyboard focus to shift to that monitor.  
Move the mouse to monitor 2 and open a window. I would expect it to open where the mouse is(monitor 2), but it does not.  
What I infer from this is that the focus is never shifted.  
Clicking on monitor 2 before opening the window opens it as expected.  
If a window is open on monitor 2 a mouse over shifts focus as expected.  
Also of interest is that when I open rofi, it opens where the mouse is (also dunst notifications but I have set it to follow my mouse in its config). So do floating windows behave differently?  
I use Arch, and qtile is from its repos.

**Comment by elParaguayo** [comment]
I think the behaviour you describe is expected:
- Windows (tiled and floating) open in the currently focused group/screen
- Screens are not focused by a mouse entering that screen
- Clicking a screen focuses it
- Rofi has an option to open on the monitor with the mouse and that is supported by Qtile

This issue was discussed a bit in #2657 too.

**Comment by m-col** [comment]
>  Also of interest is that when I open rofi, it opens where the mouse is (also dunst notifications but I have set it to follow my mouse in its config). So do floating windows behave differently?

Focussing an output is a Qtile concept, rather than X11 concept, so rofi doesn't know which one is focussed. That's why these programs will can decide to: 1. open in a set output, 2. open on the currently focussed window's output, or 3. open on the output containing the mouse.

> This issue was discussed a bit in https://github.com/qtile/qtile/issues/2657 too.

Reading through this again it looks like we (or I) was pro tracking cursor motion across outputs, but one thing missing from that issue (maybe it's in another) is that, while this is simple when moving across the root window, moving across the screen boundaries while over a window and toggling that would _also_ require tracking cursor motion for each window. This is also possible, just, you know even more cursor tracking.

Another caveat: consider dragging a window from near its left edge rightward over a screen boundary. Currently the screen focus switches when the centre of that window crosses the screen boundary, while the mouse is still on the initial screen. I'm not sure what the best approach there would be as both options might seem intuitive.

**Comment by arjan-s** [comment]
Actually this is something I've missed as well after switching from sway.
Perhaps, instead of constantly tracking mouse movements, it could be implemented as part of determining the placement of new windows? If `focus_follows_mouse = True`, then (if no other rules say otherwise) place a new window on the output where the mouse is.

**Comment by SweepingBishops** [comment]
How much does qtile track the mouse right now?  

I've read through #2657 and it seems to me that it is heavily focused on mouse tracking. Arjan-s' solution makes sense to me.  
If I was typing in a text editor and moved my cursor to the other monitor(not clicking), I would still expect my keystrokes to be captured by it. If I click on the other monitor, or if there is a window open there I would expect the focus to shift.  

How hard will it be to do something like Arajan-s said? When you open the window, just checking where the mouse is and acting according to that.  
>Another caveat: consider dragging a window from near its left edge rightward over a screen boundary. Currently the screen focus switches when the centre of that window crosses the screen boundary, while the mouse is still on the initial screen. I'm not sure what the best approach there would be as both options might seem intuitive.  

For this it seems to me that if the `focus_follows_mouse = True` then open it on the output where the mouse is, otherwise open it as it is being now.  
Is what I am suggesting doable, or am I complete nuts?

**Comment by SweepingBishops** [comment]
Hello?

**Comment by elParaguayo** [comment]
I wonder if you could create a function in your config which is run on the `client_new` hook. In that function you can query the mouse position (`qtile.core.get_mouse_position()`), work out which screen the mouse is on, find the group on that screen and add the window to that group.

**Comment by SweepingBishops** [comment]
I'll see if I can get that to work. Thanks!

**Comment by SweepingBishops** [comment]
This is what I wrote.
```
@hook.subscribe.client_new
def to_screen(client):
    mouse_xpos = qtile.core.get_mouse_position()[0]
    if mouse_xpos <= 1367:
        client.cmd_toscreen(0)
    else:
        client.cmd_toscreen(1)
```
Assuming two monitors 1366x768 on the left and 1920x1080 on the right.

I don't see any glaring mistakes in it, but it doesn't work as expected.
There are no errors or anything in the logs.
It behaves differently for each case, like focus on monitor one and mouse on another.
It sometimes Matches the window according to the rules I have sometimes not. I'm positive that there is some pattern to it, but sadly I don't have much time to devote to this.

Is this a me problem? If it is, should I mark this issue as solved? Any help would be appreciated.

**Comment by elParaguayo** [comment]
Is it any different if you use the `client_managed` hook?

**Comment by SweepingBishops** [comment]
I haven't tested it thoroughly yet, but it mostly works as expected.
I say mostly because , `client_managed` gets run only after the window has been positioned. While that is how it is supposed to be, it not useful in this case. 
Suppose I have a window open on monitor 1 and mouseover monitor 2. If I open a terminal, it first opens on monitor 1 and then moves to monitor 2. Monitor 1 is now empty.

**Comment by SweepingBishops** [comment]
Reply to @m-col in #3605. No I meant this

> I wonder if you could create a function in your config which is run on the `client_new` hook. In that function you can query the mouse position (`qtile.core.get_mouse_position()`), work out which screen the mouse is on, find the group on that screen and add the window to that group.

**Comment by m-col** [comment]
Ah, think I'm getting confused as I guess there are two issues here? Though that issue (window opening on focussed screen rather than screen with the cursor) is a symptom of the fact that screen focus doesn't follow the mouse.

It would be nice to have an implementation of the hook known, and perhaps even in the default config until screen focus does follow the mouse (if ever).

**Comment by TheTerrior** [comment]
Making the switch from BSPWM to Qtile (solves so many problems for me), but this feature is the one thing I'm really missing. Bumping to see if any progress has been made. Just having the option to enable this would be great.

**Comment by elParaguayo** [comment]
Still being worked on. However, the preferred solution needs a change to a different python library and I'm hitting some issues with it.

**Comment by dmolin** [comment]
Has there been any further progress or investigation on this issue? Just checking if this is even in scope or just completely abandoned.

**Comment by dmolin** [comment]
Hi @elParaguayo , yes I'm on X11 and thank you, it would be really appreciated if we can keep this in sight somehow. I'm sadly not proficient at all in the whole Python ecosystem otherwise I'd be happy to spend some time digging into this myself

**Comment by elParaguayo** [comment]
I had some old code for this which I couldn't submit as a PR as there were some issues with other python libraries. I've fixed those issues too and, given 0.23 requires latest version of xcffib, every user will have those fixes on their system.

I need to dig out the old code and rework it as there have been a lot of changes to qtile since I first wrote it!

**Comment by jwijenbergh** [comment]
Let's reopen this

**Comment by elParaguayo** [comment]
#4557 is an attempt at this. The idea is to change screen focus as the mouse crosses from one screen to another.
